### PR TITLE
App and secret java

### PR DESCRIPTION
**Pull request by ruth0048**

### CODE REVIEW

File: TestBCClient-Android/brainCloud/src/main/java/com/bitheads/braincloud/client/BrainCloudWrapper.java [hunk 1]
```diff
@@ -68,6 +68,11 @@
     private BrainCloudClient _client = null;
     private String _wrapperName = "";
 
+    private String _lastUrl = "";
+    private String _lastSecretKey = "";
+    private String _lastGameId = "";
+    private String _lastGameVersion = "";
```
Comment (davidstl) [inline TestBCClient-Android/brainCloud/src/main/java/com/bitheads/braincloud/client/BrainCloudWrapper.java]: Why did you add that. is this used?

File: brainCloud/main/java/com/bitheads/braincloud/client/BrainCloudClient.java [hunk 2]
```diff
@@ -151,23 +152,81 @@ public RTTComms getRTTComms() {
...
+     * @param secretKey
+     *            The app id
+     * @param appId
+     *            The map of appId to secret
+     * @param appVersion
+     *            The app version (e.g. "1.0.0").
+     */
+    public void initialize(String serverURL, String appId, String secretKey, String appVersion)
+    {
+        initializeHelper(serverURL, secretKey, appId, appVersion);
```
Comment (davidstl) [inline brainCloud/main/java/com/bitheads/braincloud/client/BrainCloudClient.java]: not a fan of `initialiezHelper`. Also why did you create a new `initialize` function that swap `appId` and ` secretKey`?

File: brainCloud/main/java/com/bitheads/braincloud/client/BrainCloudClient.java [hunk 3]
```diff
@@ -151,23 +152,81 @@ public RTTComms getRTTComms() {
...
+     * @param defaultAppId
+     *            The app id
+     * @param appIdSecretMap 
+     *            The map of appId to secret
+     * @param appVersion
+     *            The app version (e.g. "1.0.0").
+     */
+    public void initializeWithApps(String defaultAppId, Dictionary<String, String> appIdSecretMap, String appVersion)
+    {
+        initializeWithApps(DEFAULT_SERVER_URL, defaultAppId, appIdSecretMap, appVersion);
```
Comment (davidstl) [inline brainCloud/main/java/com/bitheads/braincloud/client/BrainCloudClient.java]: `DEFAULT_SERVER_URL` is this new in JAVA?

File: brainCloud/main/java/com/bitheads/braincloud/client/BrainCloudClient.java [hunk 4]
```diff
@@ -151,23 +152,81 @@ public RTTComms getRTTComms() {
...
+        _comms.initializeWithApps(serverUrl, defaultAppId, appIdSecretMap);
+
+        initialized = true;
     }
 
     /**
      * Initializes the brainCloud client with your app information. This method
      * must be called before any API method is invoked.
-     *
+     *
```
Comment (davidstl) [inline brainCloud/main/java/com/bitheads/braincloud/client/BrainCloudClient.java]: ? lol

File: brainCloud/main/java/com/bitheads/braincloud/comms/RTTComms.java [hunk 5]
```diff
@@ -124,6 +124,26 @@ public void onError(Exception e) {
     private HashMap<String, IRTTCallback> _callbacks = new HashMap<String, IRTTCallback>();
 
     private String _appId;
+    public getAppId() { return _appId; }
+
+    private Dictionary<String, String> _appIdSecretMap;
+    public getAppIdSecretMap(){ return _appIdSecretMap; }
+   
+    private String _secretKey;
+    public getSecretKey()
```
Comment (davidstl) [inline brainCloud/main/java/com/bitheads/braincloud/comms/RTTComms.java]: function with no return type?
Also revert `RTTComms` changes.

File: brainCloud/main/java/com/bitheads/braincloud/comms/BrainCloudRestClient.java [hunk 6]
```diff
@@ -135,6 +137,24 @@ public void initialize(String serverUrl, String appId, String secretKey) {
...
+                break;
+            }
+        }
+
+        initialize(serverUrl, appId, secretKey);
+
+    }
+
     public void addToQueue(ServerCall serverCall) {
         _waitingQueue.add(serverCall);
```
Comment (davidstl) [inline brainCloud/main/java/com/bitheads/braincloud/comms/BrainCloudRestClient.java:153]: Comms is missing the capture of the new appId from the server.

File: TestBCClient-Android/brainCloud/src/main/java/com/bitheads/braincloud/client/BrainCloudWrapper.java (hunk 1 above)
Comment (ruth0048) [reply TestBCClient-Android/brainCloud/src/main/java/com/bitheads/braincloud/client/BrainCloudWrapper.java]: i added that for this part, so we can keep the last data we had. It seemed to be something the other libs were doing at the time

```

    private void initializeWithApps(String url, String defaultAppId, Map<String, String> secretMap, String version, String companyName, String appName)
    {
        if(_client == null)
        {
            _client = new BrainCloudClient();
        }

        String secretKey = "MISSING";
        if(secretMap.containsKey(defaultAppId))
        {
            secretKey = secretMap.get(defaultAppId);
        }

        _lastUrl = url;
        _lastSecretKey = secretKey;
        _lastGameId = defaultAppId;
        _lastGameVersion = version;

        getClient().initializeWithApps(url, defaultAppId, secretMap, version);
    } 
```
for this funtion

Comment (davidstl) [reply TestBCClient-Android/brainCloud/src/main/java/com/bitheads/braincloud/client/BrainCloudWrapper.java]: Ya remove that. We are not using this in java. Otherwise you'll have to update them once you switchapp, etc.

File: brainCloud/main/java/com/bitheads/braincloud/comms/BrainCloudRestClient.java [hunk 7]
```diff
@@ -135,6 +139,16 @@ public void initialize(String serverUrl, String appId, String secretKey) {
         }
     }
 
+    public void initializeWithApps(String serverUrl, String appId, Map<String, String> secretMap) {
+
+        //clear the map
+        _secretMap.clear();
```
Comment (davidstl) [inline brainCloud/main/java/com/bitheads/braincloud/comms/BrainCloudRestClient.java:145]: it gets overridden no necessary to clear :P
_Ignore this comment tho, no harm here._

File: brainCloud/main/java/com/bitheads/braincloud/comms/BrainCloudRestClient.java [hunk 8]
```diff
@@ -882,6 +896,20 @@ private void handleBundle(JSONObject root) throws JSONException {
...
+                                    LogString("Current appId: " + _appId);
+                                    _appId = data.getString("switchToAppId");
+                                    LogString("Changing to appId: " + _appId);
+
+                                    _secretKey = "MISSING";
+                                    if(_secretMap.containsKey(_appId))
+                                    {
+                                        _secretKey = _secretMap.get(_appId);
+                                        LogString("found: " + _appId);
+                                        LogString("secret: " + _secretMap.get(_appId));
```
Comment (davidstl) [inline brainCloud/main/java/com/bitheads/braincloud/comms/BrainCloudRestClient.java]: Remove those secret log.
